pkg/driver/driverV2.go

```diff
@@ -0,0 +1,497 @@
+	if err != nil {
+		d.logger.Errorf("Unable to install custom runners %v", err)
+		err = errs.New(errs.GenericErrRemark.Error())
+		return err
+	}
+
+	return nil
+}
+
+func (d *driverV2) setUpDiscovery(ctx context.Context,
+	payload *core.Payload,
+	tasConfig *core.TASConfigV2,
+	oauth *core.Oauth) (*setUpResultV2, error) {
+	cacheKey := fmt.Sprintf("%s/%s/%s/%s", tasConfig.Cache.Version, payload.OrgID, payload.RepoID, tasConfig.Cache.Key)
+
+	g, errCtx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		if errG := d.CacheStore.Download(errCtx, cacheKey); errG != nil {
+			d.logger.Errorf("Unable to download cache: %v", errG)
+			errG = errs.New(errs.GenericErrRemark.Error())
+			return errG
+		}
+		return nil
+	})
+	diffExists := true
+	diff := map[string]int{}
+	g.Go(func() error {
+		diffC, errG := d.DiffManager.GetChangedFiles(errCtx, payload, oauth)
+		if errG != nil {
+			if errors.Is(errG, errs.ErrGitDiffNotFound) {
+				diffExists = false
+			} else {
+				d.logger.Errorf("Unable to identify changed files %s", errG)
+				errG = errs.New("Error occurred in fetching diff from GitHub")
+				return errG
+			}
+		}
+		diff = diffC
+		return nil
+	})
+	err := g.Wait()
+	if err != nil {
+		return nil, err
+	}
+	return &setUpResultV2{
+		cacheKey:   cacheKey,
+		diffExists: diffExists,
+		diff:       diff,
+	}, nil
+}
+
+func (d *driverV2) buildDiscoveryArgs(payload *core.Payload, tasConfig *core.TASConfigV2,
+	subModule *core.SubModule,
+	secretMap map[string]string,
+	diffExists bool,
+	diff map[string]int) core.DiscoveyArgs {
+	testPattern := subModule.Patterns
+	envMap := getEnv(payload, tasConfig, subModule)
+	modulePath := path.Join(global.RepoDir, subModule.Path)
+
+	return core.DiscoveyArgs{
+		TestPattern:    testPattern,
+		Payload:        payload,
+		EnvMap:         envMap,
+		SecretData:     secretMap,
+		TestConfigFile: subModule.ConfigFile,
+		FrameWork:      subModule.Framework,
+		SmartRun:       tasConfig.SmartRun,
+		Diff:           diff,
+		DiffExists:     diffExists,
+		CWD:            modulePath,
+	}
+}
+
+func getEnv(payload *core.Payload, tasConfig *core.TASConfigV2, subModule *core.SubModule) map[string]string {
+	var envMap map[string]string
+	if payload.EventType == core.EventPullRequest {
+		envMap = tasConfig.PreMerge.EnvMap
+	} else {
+		envMap = tasConfig.PostMerge.EnvMap
+	}
+	if envMap == nil {
+		envMap = map[string]string{}
+	}
+
+	// overwrite the existing env with more specific one
+	if subModule.Prerun != nil && subModule.Prerun.EnvMap != nil {
+		for k, v := range subModule.Prerun.EnvMap {
+			envMap[k] = v
+		}
+	}
+	if path.Join(global.RepoDir, subModule.Path) == global.RepoDir {
+		envMap[global.ModulePath] = ""
+	} else {
+		envMap[global.ModulePath] = subModule.Path
+	}
+	return envMap
+}
+
+func populateTestDiscoveryV2(testDiscoveryResult *core.DiscoveryResult, subModule *core.SubModule, tasConfig *core.TASConfigV2) {
```

Comment:
No need to add V2 here as we are already inside driverV2.
One more suggestion I think instead of having two file as driver_v1 driver_v2, we should have folders and inside those folder we should have separate files and there would be no need to add V2 in names. I'm saying this because in the future we'll have more things and at that point we might need multiple files for each driver, we can put multiple files in their respective folders

Reply:
1. Cannot remove V2 as both dirver_v1 and driver_v2 share same package name
2. I have created common NodeInstaller struct which install node version in builder file , if we move driver_v1 / v2 to there respective folder then have to import driver package , since driver_v1 is separate package , It needs to imported in builder file , this will create  `package import cycle`